### PR TITLE
Replace uv_messenger with AsyncQueue; rewrite remaining uv_* in C++

### PR DESCRIPTION
**Pull request by jfirebaugh**

### CODE REVIEW

**Comment by jfirebaugh** [comment]
@kkaefer @tmpsantos Can you review please?

File: src/mbgl/util/worker.cpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        thread.join();
+    }
+
+    queue->stop();
+}
+
+void Worker::send(std::function<void ()> work,
+                  std::function<void ()> after) {
+    MBGL_VERIFY_THREAD(tid);
+    channel.send({work, after});
```
Comment (kkaefer) [inline src/mbgl/util/worker.cpp]: `send(nullptr, nullptr);` will:

1. terminate the worker
2. crash the application because `after()` on line 8 is invalid

Comment (kkaefer) [inline src/mbgl/util/worker.cpp]: We should either `assert` both `work` and `after`, use shim functions when the functions objects are empty, or discard the call altogether.

Comment (jfirebaugh) [inline src/mbgl/util/worker.cpp]: There's a guard condition for `after` in the thread function. I'll add an assert for `work`.

**Comment by jfirebaugh** [comment]
Tests were hanging because `Worker` didn't do an `unref`/`ref` dance on the `AsyncQueue`. The queue needs to be referenced iff there is a pending work item.

**Pull request merged**